scapy/contrib/automotive/someip.py

```diff
@@ -210,226 +210,221 @@ def _bind_someip_layers():
-        ShortField("len", None),
-        ByteField("type", 0),
-        ByteField("res_hdr", 0)
+SDOPTION_CFG_TYPE = 0x01
+SDOPTION_LOADBALANCE_TYPE = 0x02
+SDOPTION_LOADBALANCE_LEN = 0x05
+SDOPTION_IP4_ENDPOINT_TYPE = 0x04
+SDOPTION_IP4_ENDPOINT_LEN = 0x0009
+SDOPTION_IP4_MCAST_TYPE = 0x14
+SDOPTION_IP4_MCAST_LEN = 0x0009
+SDOPTION_IP4_SDENDPOINT_TYPE = 0x24
+SDOPTION_IP4_SDENDPOINT_LEN = 0x0009
+SDOPTION_IP6_ENDPOINT_TYPE = 0x06
+SDOPTION_IP6_ENDPOINT_LEN = 0x0015
+SDOPTION_IP6_MCAST_TYPE = 0x16
+SDOPTION_IP6_MCAST_LEN = 0x0015
+SDOPTION_IP6_SDENDPOINT_TYPE = 0x26
+SDOPTION_IP6_SDENDPOINT_LEN = 0x0015
+
+
+def _MAKE_COMMON_SDOPTION_FIELDS_DESC(type, length=None):
+    return [
+        ShortField("len", length),
+        XByteField("type", type),
+        XByteField("res_hdr", 0)
     ]
 
 
-class _SDOption_Tail(_SDOption):
-    fields_desc = [
-        ByteField("res_tail", 0),
-        ByteEnumField("l4_proto", 0x06, {0x06: "TCP", 0x11: "UDP"}),
+def _MAKE_COMMON_IP_SDOPTION_FIELDS_DESC():
+    return [
+        XByteField("res_tail", 0),
+        ByteEnumField("l4_proto", 0x11, {0x06: "TCP", 0x11: "UDP"}),
         ShortField("port", 0)
     ]
 
 
-class _SDOption_IP4(_SDOption):
-    fields_desc = [
-        _SDOption_Header,
-        IPField("addr", "0.0.0.0"),
-        _SDOption_Tail
-    ]
-
-
-class _SDOption_IP6(_SDOption):
-    fields_desc = [
-        _SDOption_Header,
-        IP6Field("addr", "2001:cdba:0000:0000:0000:0000:3257:9652"),
-        _SDOption_Tail
-    ]
-
-
-class SDOption_Config(_SDOption):
-    LEN_OFFSET = 0x01
-
+class SDOption_Config(_SDPacketBase):
     name = "Config Option"
-    _defaults = {'type': _SDOption.CFG_TYPE}
-    fields_desc = [
-        _SDOption_Header,
-        StrField("cfg_str", "")
+    fields_desc = _MAKE_COMMON_SDOPTION_FIELDS_DESC(SDOPTION_CFG_TYPE) + [
+        StrLenField("cfg_str", "\x00", length_from=lambda pkt: pkt.len - 1)
     ]
 
     def post_build(self, pkt, pay):
-        length = self.len
-        if length is None:
-            length = len(self.cfg_str) + self.LEN_OFFSET
+        if self.len is None:
+            length = len(self.cfg_str) + 1  # res_hdr field takes 1 byte
             pkt = struct.pack("!H", length) + pkt[2:]
         return pkt + pay
 
-
-class SDOption_LoadBalance(_SDOption):
+    @staticmethod
+    def make_string(data):
+        # Build a valid null-terminated configuration string from a dict or a
+        # list with key-value pairs.
+        # Example:
+        #    >>> SDOption_Config.make_string({ "hello": "world" })
+        #    b'\x0bhello=world\x00'
+        #    >>> SDOption_Config.make_string([
+        #    ...     ("x", "y"),
+        #    ...     ("abc", "def"),
+        #    ...     ("123", "456")
+        #    ... ])
+        #    b'\x03x=y\x07abc=def\x07123=456\x00'
+
+        if isinstance(data, dict):
+            data = list(data.items())
+
+        # combine entries
+        data = ["{}={}".format(k, v) for (k, v) in data]
```

Comment:
`for k,v` instead of `for (k,v)`.
Also you should use generators instead of lists whenever possible as a general practice (use `(x for in [...])` instead of `[x for x in [...]]`  )

Reply:
makes sense, I didn't know this was possible actually